core/rfb.js

```diff
@@ -1496,6 +1499,76 @@ export default class RFB extends EventTargetMixin {
         return true;
     }
 
+    _stringToAsciiByteArray(str) {
+        let bytes = new Uint8Array(str.length);
+        for (let i = 0; i < str.length; ++i) {
+            bytes[i] = str.charCodeAt(i);
+        }
+        return bytes;
+    }
+
+    _negotiateARDAuth() {
+
+        if (this._rfbCredentials.username === undefined ||
+            this._rfbCredentials.password === undefined) {
+            this.dispatchEvent(new CustomEvent(
+                "credentialsrequired",
+                { detail: { types: ["username", "password"] } }));
+            return false;
+        }
+
+        if (this._sock.rQwait("read ard", 4)) { return false; }
+
+        let generator = this._sock.rQshiftBytes(2);   // DH base generator value
+
+        let keyLength = this._sock.rQshift16();
+
+        if (this._sock.rQwait("read ard keylength", keyLength*2, 4)) { return false; }
+
+        let primeBytes = this._sock.rQshiftBytes(keyLength);  // predetermined prime modulus
+
+        let serverPubKeyBytes = this._sock.rQshiftBytes(keyLength); // other party's public key
+
+        let prime = new forge.jsbn.BigInteger(forge.util.createBuffer(primeBytes).toHex(), 16);
+
+        let clientPrivKey = new forge.jsbn.BigInteger(forge.util.createBuffer(forge.random.generate(keyLength)).toHex(), 16);
+
+        let ghd = new forge.jsbn.BigInteger(generator);
+        let clientPubKey = ghd.modPow(clientPrivKey, prime);
+
+        let serverPubKey = new forge.jsbn.BigInteger(forge.util.createBuffer(serverPubKeyBytes).toHex(), 16);
+
+        let sharedKey = serverPubKey.modPow(clientPrivKey, prime);
+
+        let md5Shared = forge.md.md5.create().update(forge.util.hexToBytes(sharedKey.toString(16))).digest();
+
+        let userPassword = this._rfbCredentials.username + '\0'
+                + forge.util.createBuffer(forge.random.generate(32)).toHex().substring(0, 63-this._rfbCredentials.username.length) +
+                this._rfbCredentials.password + '\0'
+                + forge.util.createBuffer(forge.random.generate(32)).toHex().substring(0, 63-this._rfbCredentials.password.length);
+
+        let packedUserPassword = forge.util.encodeUtf8(userPassword);
```

Comment:
This seems odd. Are you sure this shouldn't be done to the fields before they are concatenated? Right now the total structure will have varying length, and you never send the length to the server.

Reply:
The packed user password is always, 128 bytes in length, but you are right if someone who has a user or password that is larger than 63 characters this would not work, I can change the code to handle that case.

Reply:
My point was that encoding it to UTF8 **expands** the data. A 128 character string might end up being 140 bytes once you've encoded it. You should do the encoding as the **first** step, not the last.